File: LeetCode/1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.py
from sortedcontainers import SortedList

class Solution:
    def longestSubarray(self, nums: List[int], limit: int) -> int:
        n = len(nums)
        st = SortedList()
        i = 0
        j = 0
        maxLength = 0

        while j < n:
            st.add(nums[j])

            while st[-1] - st[0] > limit:
                st.remove(nums[i])
                i += 1

            maxLength = max(maxLength, j - i + 1)
            j += 1

        return maxLength




# class Solution:
#     def longestSubarray(self, nums: List[int], limit: int) -> int:
#         maxpq = []
#         minpq = []
#         i, j = 0, 0
#         maxlen = 0
#         n = len(nums)

#         while j < n:
#             heappush(maxpq, (-nums[j], j))
#             heappush(minpq, (nums[j], j))

#             while -maxpq[0][0] - minpq[0][0] > limit:
#                 i = min(maxpq[0][1], minpq[0][1]) + 1

#                 while maxpq and maxpq[0][1] < i:
#                     heappop(maxpq)
#                 while minpq and minpq[0][1] < i:
#                     heappop(minpq)
                
#             maxlen = max(maxlen, j - i + 1)
#             j += 1
        
#         return maxlen

File: LeetCode/1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.java
class Solution {
    public int longestSubarray(int[] nums, int limit) {
        int n = nums.length;
        TreeMap<Integer, Integer> mp = new TreeMap<>();
        int i = 0, j = 0, maxlen = 0;
        while (j < n) {
            mp.put(nums[j], mp.getOrDefault(nums[j], 0) + 1);
            while (mp.lastKey() - mp.firstKey() > limit) {
                mp.put(nums[i], mp.get(nums[i]) - 1);
                if (mp.get(nums[i]) == 0) {
                    mp.remove(nums[i]);
                }
                i++;
            }
            maxlen = Math.max(maxlen, j - i + 1);
            j++;
        }
        return maxlen;
    }
}




// class Solution {
//     public int longestSubarray(int[] nums, int limit) {
//         int n = nums.length;
//         PriorityQueue<int[]> mpq = new PriorityQueue<>((a, b) -> b[0] - a[0]);
//         PriorityQueue<int[]> lpq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
//         int i = 0, j = 0;
//         int maxlen = 0;
//         while (j < n) {
//             mpq.offer(new int[]{nums[j], j});
//             lpq.offer(new int[]{nums[j], j});
//             while (mpq.peek()[0] - lpq.peek()[0] > limit) {
//                 i = Math.min(mpq.peek()[1], lpq.peek()[1]) + 1;
//                 while (mpq.peek()[1] < i) {
//                     mpq.poll();
//                 }
//                 while (lpq.peek()[1] < i) {
//                     lpq.poll();
//                 }
//             }
//             maxlen = Math.max(maxlen, j - i + 1);
//             j++;
//         }
//         return maxlen;
//     }
// }